CRM: PHP notices: Translation loading for the zero-bs-crm domain was triggered too early. #3542

mikestottuk opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes Automattic/zero-bs-crm#3542

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Removed the plugins_loaded hook registration that was calling load_textdomain() from the init_hooks() method in ZeroBSCRM.Core.php
* This prevents the text domain from being loaded too early in the WordPress initialization sequence
* Fixes the PHP notice: "Function _load_textdomain_just_in_time was called incorrectly. Translation loading for the zero-bs-crm domain was triggered too early."

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Enable WP_DEBUG and WP_DEBUG_LOG in wp-config.php
* Activate the plugin
* Check debug.log for any "translation loaded too early" notices
* Verify that all translations still work correctly in the admin area
* Test with different languages to ensure translations are loading properly
* Verify that no other functionality is affected by this change

mikestottuk committed:
new activation redirect and translation loading

mikestottuk committed:
Changelog

mikestottuk committed:
removing translation loading from class

mikestottuk committed:
removing translation loading from class

mikestottuk committed:
reapply core changes

mikestottuk committed:
fix the bootstrap test

mikestottuk committed:
manually defining the test again

mikestottuk committed:
show wizard if activating mulitple plugins - for unit test testing

mikestottuk committed:
wizard and tests

mikestottuk committed:
improving activation tests

mikestottuk committed:
firming up tests

mikestottuk committed:
change test to allow for redirection

mikestottuk committed:
more tests

mikestottuk committed:
fixing the bulk activation test for wizard not shown

mikestottuk committed:
fixing plugin name in activation test

mikestottuk committed:
fixing plugin name in activation test

mikestottuk committed:
making it so the jpcrm_database_engine does not rely on

mikestottuk committed:
removing wait

mikestottuk committed:
no wizard on upgrade

mikestottuk committed:
upgrade test

tbradsha committed:
Skip feedback on bulk deactivation

tbradsha committed:
Reshuffle activation redirect logic

tbradsha commented on projects/plugins/crm/ZeroBSCRM.php:
```diff
@@ -324,9 +383,26 @@ function zeroBSCRM_init_perfTest() {
-	$zbs = ZeroBSCRM::instance();
+	// init hook
+	add_action(
+		'init',
```
I'll note that we're changing the load order here; I'm not sure what implications this might have for extensions that register with CRM (e.g. Mail Campaigns or WooSync).

On the plus side, this sets the stage to clean up several lines in `ZeroBSCRM::init_hooks()` to no longer use the `init` hook.

tbradsha commented on projects/plugins/crm/includes/ZeroBSCRM.Core.php:
```diff
@@ -1353,20 +1356,9 @@ public function includes() {
 		do_action( 'before_zerobscrm_init' );
 
-		// After all the plugins have loaded (THESE FIRE BEFORE INIT)
-		add_action( 'plugins_loaded', array( $this, 'load_textdomain' ) ); // } Translations
```
Should we chop `ZeroBSCRM::load_textdomain` now?

mikestottuk replied:
we should!

tbradsha commented on projects/plugins/crm/includes/ZeroBSCRM.Database.php:
```diff
@@ -951,10 +951,23 @@ function zeroBSCRM_DB_canInnoDB(){
-		return $zbs->database_server_info['db_engine_label'];
+
+	// If $zbs is an object but database_server_info is not set, try to populate it.
+	if ( is_object( $zbs ) && ! isset( $zbs->database_server_info ) ) {
```
Did you run into an issue with this? If `$zbs` is not an object, we should probably return early.

Related, `$zbs->database_server_info` will always be set if `$zbs` is an object, given that it's initialized as an empty object. It's also populated as early as possible (within `$zbs->verify_minimum_requirements()`).

In other words, if `$zbs` is an object, the rest of the checks are unnecessary and the fallback scenario isn't possible.

tbradsha commented on projects/plugins/crm/tests/codeception/acceptance/11_JPCRM_Activation_Cest.php:
```diff
@@ -26,22 +26,50 @@ public function catch_wp_update_pages( AcceptanceTester $I ) {
 
-		// Activating the plugin directly loads the welcome wizard, so no need to move pages here.
+		// Click the activate link for jetpack-crm directly
+		$I->click( 'Activate', array( 'css' => 'tr[data-slug="jetpack-crm"] .activate a' ) );
```
I presume this is because the `$I->activatePlugin()` uses the bulk activation route? This might be irrelevant now that I've adjusted it to support bulk activation with one plugin.

tbradsha commented on projects/plugins/crm/tests/php/bootstrap.php:
```diff
@@ -79,12 +79,22 @@
 	global $zbs;
-	$zbs->install();
-	zeroBSCRM_notifyme_createDBtable();
+	if ( class_exists( 'zeroBSCRM' ) && ! isset( $zbs ) ) {
```
Why do we need this check? We literally load the class file and we know it's not set. :^)

tbradsha commented on projects/plugins/crm/tests/php/bootstrap.php:
```diff
@@ -79,12 +79,22 @@
+	// Run installation if needed
+	if ( isset( $zbs ) && is_object( $zbs ) && method_exists( $zbs, 'install' ) ) {
+		$zbs->install();
+		if ( function_exists( 'zeroBSCRM_notifyme_createDBtable' ) ) {
```
Same thing here...if `$zbs` is instantiated, that means it's run the `__construct()`, which means it's run `$zbs->includes()`, which means this function does exist. There's no way to get to this point and not have the function exist.

tbradsha commented on projects/plugins/crm/tests/php/bootstrap.php:
```diff
@@ -79,12 +79,22 @@
+	}
+
+	// Run installation if needed
+	if ( isset( $zbs ) && is_object( $zbs ) && method_exists( $zbs, 'install' ) ) {
```
This is overkill and not needed. :^)

mikestottuk replied:
unless someone deletes the install function?

tbradsha reviewed (commented):
Some comments to consider...

tbradsha commented on projects/plugins/crm/tests/codeception/acceptance/11_JPCRM_Activation_Cest.php:
```diff
@@ -26,22 +26,50 @@ public function catch_wp_update_pages( AcceptanceTester $I ) {
+		$this->assertWizardIsShown( $I );
+	}
+
+	public function test_bulk_plugin_activation_skips_wizard( AcceptanceTester $I ) {
```
I suspect this doesn't show the wizard because technically the option already is set. So more accurately this is testing if the wizard shows on second load.

tbradsha replied:
d2e27d5

tbradsha committed:
Merge branch 'trunk' into fix/crm/clean-translation-loading

mikestottuk committed:
some tidy

mikestottuk committed:
more tidy

mikestottuk committed:
tidy acceptance

mikestottuk committed:
tidy acceptance with jetpack-crm slug

tbradsha commented on projects/plugins/crm/includes/ZeroBSCRM.Core.php:
```diff
@@ -2262,37 +2262,6 @@ private function is_active_theme( $theme ) {
 	public function frontend_includes() {
 	}
 
-	/**
```
🪓

tbradsha committed:
Remove unneeded check

tbradsha committed:
Merge branch 'trunk' into fix/crm/clean-translation-loading

tbradsha committed:
Chop bulk activation wizard skip test

This won't fire regardless since the `jpcrm_wizard_completed` option is already set at this point.

Also, we now only skip the wizard if more than one plugin is activated via this method.

tbradsha committed:
Another try

tbradsha committed:
Fine, let's go with this
